torch_xla/csrc/ops/einsum_utilities.h

```diff
@@ -0,0 +1,75 @@
+#include <algorithm>
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace torch_xla {
+
+class EinsumUtilities {
+ public:
+  static std::vector<std::string> BuildBackwardsEquations(
+      const std::string& equation) {
+    std::vector<std::string> elements = ParseEquation(equation);
+    std::vector<std::string> equations;
+    equations.push_back(elements[2] + "," + elements[1] + "->" + elements[0]);
+    equations.push_back(elements[0] + "," + elements[2] + "->" + elements[1]);
+    return equations;
+  }
+
+  static std::string BuildBackwardsEquation(const std::string& equation) {
+    int split_index = equation.find("->");
+
+    std::string equation_input = equation.substr(0, split_index);
+    std::string equation_output =
+        equation.substr(split_index + 2, equation.size() - split_index - 2);
+
+    std::string backward_equation = equation_output + "->" + equation_input;
+    return backward_equation;
+  }
+
+  // An einsum equation is invalid if there are indices in one of the inputs or
+  // output which are not in any other input or output. This is because such
+  // equations lead to a failure when attempting to execute einsum backward on
+  // XLA.
+  static bool EquationIsValid(const std::string& equation) {
+    std::vector<std::string> elements = ParseEquation(equation);
+    for (size_t i = 0; i < elements.size(); i++) {
+      for (char c : elements[i]) {
+        // We use j to skip searching the element for its own characters
+        size_t j = 0;
+        if (std::all_of(elements.cbegin(), elements.cend(),
+                        [&c, &i, &j](std::string element) {
+                          return j++ == i ||
+                                 std::find(element.begin(), element.end(), c) ==
+                                     element.end();
+                        })) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+ private:
+  static std::vector<std::string> ParseEquation(const std::string& equation) {
```

Comment:
can you give example `equation` and example output? String parsing logic is always hard to read.

Reply:
Done